src/absil/ilwritepdb.fs

```diff
@@ -0,0 +1,634 @@
+    ()
+
+let GetRowCounts tableRowCounts =
+    let builder = ImmutableArray.CreateBuilder<int>(tableRowCounts |> Array.length)
+    tableRowCounts |> Seq.iter(fun x -> builder.Add(x))
+    builder.MoveToImmutable()
+
+let fixupOverlappingSequencePoints fixupSPs showTimes methods =
+    // This next bit is a workaround.  The sequence points we get 
+    // from F# (which has nothing to do with this module) are actually expression 
+    // marks, i.e. the source ranges they denote are typically 
+    // nested, and each point indicates where the  
+    // code for an expression with a particular range begins.  
+    // This is in many ways a much more convenient form to emit. 
+    // However, it is not the form that debug tools accept nicely. 
+    // However, sequence points are really a non-overlapping, non-nested 
+    // partition of the source code of a method.  So here we shorten the 
+    // length of all sequence point marks so they do not go further than 
+    // the next sequence point in the source. 
+    let spCounts =  methods |> Array.map (fun x -> x.SequencePoints.Length)
+    let allSps = Array.concat (Array.map (fun x -> x.SequencePoints) methods |> Array.toList)
+    let allSps = Array.mapi (fun i sp -> (i,sp)) allSps
+    if fixupSPs then 
+        // sort the sequence points into source order 
+        Array.sortInPlaceWith (fun (_,sp1) (_,sp2) -> SequencePoint.orderBySource sp1 sp2) allSps
+        // shorten the ranges of any that overlap with following sequence points 
+        // sort the sequence points back into offset order 
+        for i = 0 to Array.length allSps - 2 do
+            let n,sp1 = allSps.[i]
+            let _,sp2 = allSps.[i+1]
+            if (sp1.Document = sp2.Document) && 
+               (sp1.EndLine > sp2.Line || 
+                (sp1.EndLine = sp2.Line &&
+                 sp1.EndColumn >= sp2.Column)) then
+              let adjustToPrevLine = (sp1.Line < sp2.Line)
+              allSps.[i] <-  n,{sp1 with EndLine = (if adjustToPrevLine then sp2.Line-1 else sp2.Line)
+                                         EndColumn = (if adjustToPrevLine then 80 else sp2.Column) }
+        reportTime showTimes (sprintf "PDB: fixupOverlappingSequencePoints %d" (allSps |> Array.length) )
+        Array.sortInPlaceBy fst allSps
+    spCounts, allSps
+
+let WritePortablePdbInfo (fixupSPs:bool) showTimes fpdb (info:PdbData) = 
+    try FileSystem.FileDelete fpdb with _ -> ()
+    sortMethods showTimes info
+    let _spCounts, _allSps = fixupOverlappingSequencePoints fixupSPs showTimes info.Methods
+    let externalRowCounts = GetRowCounts info.TableRowCounts
+    let docs = 
+        if info.Documents = null then 
+            Array.empty<PdbDocumentData>
+        else
+            info.Documents
+ 
+    let metadata = MetadataBuilder()
+    let serializeDocumentName (name:string) =
+        let count s c = s |> Seq.filter(fun ch -> if c = ch then true else false) |> Seq.length
```

Comment:
I see what your saying, but it will be a long time before I can read that idiom correctly without prompting.